python/looker_sdk/rtl/model.py

```diff
@@ -34,6 +43,124 @@ class Model:
     """Base model for all generated models.
     """
 
+    def _key_to_attr(self, key):
+        """Appends the trailing _ to python reserved words.
+        """
+        if key[-1] == "_":
+            raise KeyError(key)
+        if key in keyword.kwlist:
+            key = f"{key}_"
+        return key
+
+    def __getitem__(self, key):
+        key = self._key_to_attr(key)
+        try:
+            ret = getattr(self, key)
+        except AttributeError:
+            raise KeyError(key)
+
+        if isinstance(ret, enum.Enum):
+            ret = ret.value
+        if ret is None:
+            raise KeyError(key)
+        return ret
+
+    def __setitem__(self, key, value):
+        key = self._key_to_attr(key)
+        if not hasattr(self, key):
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{key}'"
+            )
+        annotation = self.__annotations__[key]
+        if isinstance(annotation, ForwardRef):
+            actual_type = eval(
+                annotation.__forward_arg__, self.__global_context, locals()
+            )
+            if isinstance(actual_type, enum.EnumMeta):
+
+                # untyped because mypy really doesn't like this enum internals stuff
+                def err(val):
+                    valid = []
+                    for v in actual_type.__members__.values():
+                        if v.value != "invalid_api_enum_value":
+                            valid.append(v.value)
+                    return (
+                        f"Invalid value '{val}' for "  # type: ignore
+                        f"'{self.__class__.__name__}.{key}'. Valid values are "
+                        f"{valid}"  # type: ignore
+                    )
+
+                if isinstance(value, actual_type):
+                    raise ValueError(err(value))
+                enum_member = actual_type(value)
+                if enum_member.value == "invalid_api_enum_value":
+                    raise ValueError(err(value))
+                value = enum_member
+            elif issubclass(actual_type, Model):
+                value = cattr.structure(value, actual_type)
+
+        return setattr(self, key, value)
+
+    def __delitem__(self, key):
+        self[key]  # validates key
+        setattr(self, self._key_to_attr(key), None)
+
+    def __iter__(self):
+        return iter(cattr.unstructure(self))
+
+    def __len__(self):
+        return len(cattr.unstructure(self))
+
+    def __contains__(self, key):
+        return key in cattr.unstructure(self)
+
+    def keys(self):
+        return cattr.unstructure(self).keys()
+
+    def items(self):
+        return cattr.unstructure(self).items()
+
+    def values(self):
+        return cattr.unstructure(self).values()
```

Comment:
can/should `cattr.unstructure(self)` be evaluated once, or must it be evaluated every time?

Reply:
I struggled with this a little bit - my mind went right to "how inefficient!" - My next thought was

> oh man, trying to cache this stuff and keep it in sync as they manipulate the "dictionary" is going to be really tedious and tricky

So I decided to opt for easy correctness for now: calling these methods will dump the data at that moment so it will be correct. I figure if people notice/complain about performance hits using dictionary mode then we can start to profile and figure it out. But probably not worth the extra investment without getting feedback first that people even use this stuff.

Reply:
Good plan